### PR TITLE
HPCC-18854 Roxie fails with "STATE: activity 11 reset without stop

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
Tested the previously failing case manually. Smoketest should (hopefully) catch any regressions

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday

**Comment by richardkchapman** [comment]
@ghalliday - This may result in some stops on activities that are already stopped, but that's ok.

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -2755,6 +2755,15 @@ class CRoxieServerMultiInputBaseActivity : public CRoxieServerActivity
             return NULL;
     }
 
+    virtual void stop() override
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp:2758]: @richardkchapman 
Some of the base classes already have some code like this e.g.
NonEmpty, Case - conditionally depending on what has been called already

As long as it is ok to call stop() multiple times then it shouldn't make things worse, but not convinced it is the "coorect" fix.  Worth looking a bit more?

Comment (richardkchapman) [reply roxie/ccd/ccdserver.cpp:2758]: I did spend a little while looking at whether it was necessary to check that we don't call stop more than once. In general I don't think it's possible to be 100% sure that you do not do so, nor is it necessary given there is code in several places explicitly to cope with having called stop more than once.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**